### ISSUE TITLE
clarification required on expected sign response in u2f tests

### ISSUE THREAD

**Issue by ntajunnisha** [google/u2f-ref-code#138]
In U2FTest.cc, in the below mentioned portion of code
`PASS(ctr1 = test_Sign(0x9000));
  **PASS(test_Sign(0x6985));**`
  
For the second sign call(Line No:366) after ctr1 which is hightlighted, the sign response is expected to fail with 6985 code.
But there is logical explanation/justification that why is that expected to fail for a valid sign request without any reason.

**Comment by apowers313** [comment]
Technically 0x6985 (SW_CONDITIONS_NOT_SATISFIED) isn't a failure. U2F 1.0 Section 3.3 states that it means `The request was rejected due to test-of-user-presence being required`. I'd have to go back and look at that specific test to see if the test is doing something like a `check-only` command (U2F 1.0 Section 5.1) to ensure that this is the appropriate behavior.

**Comment by ntajunnisha** [comment]
This is not a check-only sign call as check-only calls have second parameter as true. Above one is regular sign call and it sends valid sign request with that call.

**Comment by ntajunnisha** [comment]
By failure we mean that is not a success status(9000).

**Comment by apowers313** [comment]
I assume you're referring to [this test](https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L361-L371)?

``` js
  // Sign with check only should not produce signature.
  PASS(test_Sign(0x6985, true));

  uint32_t ctr1;
  PASS(ctr1 = test_Sign(0x9000));
  PASS(test_Sign(0x6985));

  WaitForUserPresence(device, arg_hasButton);

  uint32_t ctr2;
  PASS(ctr2 = test_Sign(0x9000));
```

I see your point about `test_Sign` should have `true` as a second arg. Maybe @jehrensvard could weigh in?

**Comment by ntajunnisha** [comment]
Yes, Only the line that is shown in BOLD in the same test below. 
`// Sign with check only should not produce signature.
  PASS(test_Sign(0x6985, true));

  uint32_t ctr1;
  PASS(ctr1 = test_Sign(0x9000));
  **PASS(test_Sign(0x6985));**

  WaitForUserPresence(device, arg_hasButton);

  uint32_t ctr2;
  PASS(ctr2 = test_Sign(0x9000));`

**Comment by juanlang** [comment]
As I understand it, the line you bolded is intending to check that only ONE signature is released for a positive user presence test. The next sign call immediately thereafter should not pass the user presence test.

**Comment by apowers313** [comment]
I do see that [the first TUP is supposed to consume the touch](https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L341-L344), and then `WaitForUserPresence` [re-initializes the device](https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L56-L61). But I have a hard time correlating that back to the specification -- there doesn't seem to be anything that says that a device has to reset after a touch or that only one touch is allowed per `U2FHID_INIT` in either the raw protocol or HID specs.

**Comment by mschilder123** [comment]
On Tue, Nov 22, 2016 at 11:40 AM, Adam Powers <notifications@github.com>
wrote:

> I do see that the first TUP is supposed to consume the touch
> <https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L341-L344>,
> and then WaitForUserPresence re-initializes the device
> <https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L56-L61>.
> But I have a hard time correlating that back to the specification -- there
> doesn't seem to be anything that says that a device has to reset after a
> touch or that only one touch is allowed per U2FHID_INIT in either the raw
> protocol or HID specs.
>
The test indeed aims to check that the user presence "token" got consumed..
thus an immediate 2nd sign request is expected to yield 6985 (i.e. "single
signature per touch"). It is about the reset of user presence state more
than anything related to the HID layer.

**Comment by apowers313** [comment]
> thus an immediate 2nd sign request is expected to yield 6985 (i.e. "single
signature per touch")

Thanks. Where does the spec say that?

**Comment by mschilder123** [comment]
On Wed, Nov 23, 2016 at 12:55 AM, Adam Powers <notifications@github.com>
wrote:

> thus an immediate 2nd sign request is expected to yield 6985 (i.e. "single
> signature per touch")
>
> Thanks. Where does the spec say that?
>
This test predates the spec. Perhaps the spec forgot to mention "one
signature per touch" in explicit terms?

**Comment by juanlang** [comment]
See also the FIDO U2F Overview [1]:

"A user has to activate the U2F device (i.e., 'press the button') before it will issue a key pair (for registration) or sign a challenge."

While this does not explicitly state that a single touch can only yield a single key pair or signature, it certainly implies it.

1: https://fidoalliance.org/specs/fido-u2f-v1.0-nfc-bt-amendment-20150514/fido-u2f-overview.html

**Comment by apowers313** [comment]
Thanks guys, that really helps.

It looks like the requirement is that the authnr be re-initialized after each TUP. If not re-initialized, it will return an error of SW_CONDITIONS_NOT_SATISFIED. 

Is that right?

**Comment by ranjitnarjala** [comment]
As has been mentioned earlier in this thread, there doesn't seem to be anything in the spec that requires the device to be reset after a touch. Why cannot the device still ask for a user presence verification (touch) on a second sign request, without having to first be re-initialized?

**Comment by tajnisha** [comment]
The test tool determine "one signature per touch" based on the boolean flag derived from "-b" argument, in
https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L343-L344
and
https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L336-L337

But this is currently not done in https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L366-L366
Seems the test tool can be updated in this line to invoke sign based on the boolean flag.

**Comment by ntajunnisha** [comment]
The test tool determine "one signature per touch" based on the boolean flag derived from "-b" argument, in
https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L343-L344
and
https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L336-L337

But this is currently not done in https://github.com/fido-alliance/google-u2f-ref-code/blob/792ae5d3e289abbae4d4414f39d9bffa4310fef3/u2f-tests/HID/U2FTest.cc#L366-L366
Looks like the test tool can be updated in this line to invoke sign based on the boolean flag.

**Comment by harshaintc** [comment]
Also the same assumption does not hold good with the rest of the calls.

See below.
`PASS(test_Enroll(0x9000));

  // Fob with button should have consumed touch.
  if (arg_hasButton) PASS(test_Sign(0x6985));

  // Sign with check only should not produce signature.
  PASS(test_Sign(0x6985, true));

  // Sign with wrong hk.
  regRsp.keyHandleCertSig[0] ^= 0x55;
  PASS(test_Sign(0x6a80));
  regRsp.keyHandleCertSig[0] ^= 0x55;

  // Sign with wrong appid.
  regReq.appId[0] ^= 0xaa;
  PASS(test_Sign(0x6a80));
  regReq.appId[0] ^= 0xaa;
`

Starting from https://github.com/google/u2f-ref-code/blob/master/u2f-tests/HID/U2FTest.cc#L350.

The PASS(test_Enroll(0x9000)); call should logically consume the touch and the next sign call except the auth check only should call the u2f_init. 

But for the  // Sign with wrong hk. test case the u2f_init is not called and yet the test case expects wrong data error code 0x6a80.
